# simulation

requirments for simulation

```bash
pip install pyfaidx pysam rich fire pandas 'seaborn>=0.11.2'
```

```bash
>python simulation.py -h

NAME
    simulation.py - generate testing files based on gtf and fasta files.

SYNOPSIS
    simulation.py GTF FASTA <flags>

DESCRIPTION
    Keyword arguments:
    gtf -- path to reference gtf file
    fasta -- path to genome fasta file
    bed -- path to output tss.bed. default: ./tss.bed
    bam -- path to output simulated bam file. default: ./simu.bam
    total -- how many transcripts used to generate simulated data. default: 10000
    n_jobs -- how many processes to use. default: 10

POSITIONAL ARGUMENTS
    GTF
    FASTA

FLAGS
    --bed=BED
    --bam=BAM
    --total=TOTAL
    --n_jobs=N_JOBS

NOTES
    You can also use flags syntax for POSITIONAL ARGUMENTS
```

```bash
NAME
    distance.py - stats the distance between inferred sites and simulated sites.

SYNOPSIS
    distance.py TSS OUTPUT [FILES]...

DESCRIPTION
    Keyword arguments:
    tss -- path to tss.bed generated by simulation.py
    output -- path to output directory
    files -- path to result files by ats model.

POSITIONAL ARGUMENTS
    TSS
    OUTPUT
    FILES

NOTES
    You can also use flags syntax for POSITIONAL ARGUMENTS
```
